File: next/dweb-browser/Helper/src/ChangeableSet.cs
namespace DwebBrowser.Helper;

public class ChangeableSet<E> : HashSet<E>
{
    private readonly LazyBox<Changeable<ChangeableSet<E>>> LazyChangeable = new();
    private Changeable<ChangeableSet<E>> Changeable => LazyChangeable.GetOrPut(() => new Changeable<ChangeableSet<E>>(this));

    public void OnChangeAdd(Signal<ChangeableSet<E>> cb)
    {
        Changeable.Listener.OnListener += cb;
    }

    public void OnChangeRemove(Signal<ChangeableSet<E>> cb)
    {
        Changeable.Listener.OnListener -= cb;
    }

    public void OnChangeEmit()
    {
        Changeable.EmitChange();
    }

    public new bool Add(E element)
    {
        return base.Add(element).Also(it =>
        {
            if (it)
            {
                OnChangeEmit();
            }
        });
    }

    public new void Clear()
    {
        base.Clear();
        OnChangeEmit();
    }

    public new bool Remove(E element)
    {
        return base.Remove(element).Also(it =>
        {
            if (it)
            {
                OnChangeEmit();
            }
        });
    }
}

File: next/kmp/helper/src/commonMain/kotlin/org/dweb_browser/helper/ChangeableSet.kt
package org.dweb_browser.helper

import kotlin.coroutines.CoroutineContext


public class ChangeableSet<E>(
  public val source: MutableSet<E> = mutableSetOf(),
  context: CoroutineContext = defaultAsyncExceptionHandler,
) :
  MutableSet<E> by source {
  private val changeable = Changeable(this, context)
  public fun setContext(context: CoroutineContext) {
    changeable.context = context
  }

  public val onChange: Signal.Listener<ChangeableSet<E>> get() = changeable.onChange

  public suspend fun emitChange(): Unit = changeable.emitChange()

  override fun add(element: E): Boolean =
    source.add(element).also { if (it) changeable.emitChangeBackground() }

  override fun clear(): Unit = source.clear().also { changeable.emitChangeBackground() }
  override fun remove(element: E): Boolean =
    source.remove(element).also { if (it) changeable.emitChangeBackground() }

  override fun addAll(elements: Collection<E>): Boolean =
    source.addAll(elements).also { if (it) changeable.emitChangeBackground() }

  override fun removeAll(elements: Collection<E>): Boolean =
    source.removeAll(elements).also { if (it) changeable.emitChangeBackground() }

  override fun retainAll(elements: Collection<E>): Boolean =
    source.retainAll(elements).also { if (it) changeable.emitChangeBackground() }

  /** 重置 清空所有的事件监听，清空所有的数据  */
  public fun reset() {
    changeable.clear()
    this.clear()
  }
}